tfhe/src/high_level_api/tag.rs

```diff
@@ -0,0 +1,331 @@
+    }
+
+    pub fn set_data(&mut self, data: &[u8]) {
+        match self {
+            Self::Stack { bytes, len } => {
+                if data.len() > bytes.len() {
+                    // There is not enough space, so we have to allocate
+                    // a Vec
+                    *self = Self::Heap(data.to_vec());
+                } else {
+                    bytes[..data.len()].copy_from_slice(data);
+                    *len = data.len() as u8;
+                }
+            }
+            Self::Heap(vec) => {
+                // Even if the data could fit in the Stack array,
+                // Since, we already have a vec allocated we use it instead.
+                //
+                // And in that case, there won't be any allocations since,
+                // to have a vec in the first place, the allocated size is >
+                // size_of::<Vec<T>>
+                vec.clear();
+                vec.extend_from_slice(data);
+            }
+        }
+    }
+
+    pub fn set_u64(&mut self, value: u64) {
+        let le_bytes = value.to_le_bytes();
+        self.set_data(le_bytes.as_slice());
+    }
+
+    pub fn set_u128(&mut self, value: u128) {
+        let le_bytes = value.to_le_bytes();
+        self.set_data(le_bytes.as_slice());
+    }
+}
+
+impl serde::Serialize for SmallVec {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_bytes(self.data())
+    }
+}
+
+struct SmallVecVisitor;
+
+impl<'de> serde::de::Visitor<'de> for SmallVecVisitor {
+    type Value = SmallVec;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("a slice of bytes (&[u8]) or Vec<u8>")
+    }
+
+    fn visit_bytes<E>(self, bytes: &[u8]) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        let mut vec = SmallVec::default();
+        vec.set_data(bytes);
+        Ok(vec)
+    }
+
+    fn visit_byte_buf<E>(self, bytes: Vec<u8>) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(SmallVec::Heap(bytes))
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for SmallVec {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_bytes(SmallVecVisitor)
+    }
+}
+
+/// Tag
+///
+/// The `Tag` allows to store bytes alongside of entities (keys, and ciphertext)
+/// the main purpose of this system is to `tag` / identify ciphertext with their keys.
+///
+/// * When encrypted, a ciphertext gets the tag of the key used to encrypt it.
+/// * Ciphertexts resulting from operations (add, sub, etc.) get the tag from the ServerKey used
+/// * PublicKey gets its tag from the ClientKey that was used to create it
+/// * ServerKey gets its tag from the ClientKey that was used to create it
+///
+/// User can change the tag of any entities at any point.
```

Comment:
Maybe we can document that tag compatibility is not checked by tfhe-rs and that it is up to the user to do so. And that by default the tag is empty (and an example on how to set it on a freshly generated ClientKey ?)

Reply:
Yep, we might need to add that into the actual docs but I don't know if we want to do that yet

Reply:
I don't think we should say anything about compatibility ?

here the tag is just a piece of added metadata, the BC will happen to check it

Reply:
I think that's what he proposed to say in the doc, that the tag is just some data storage and tfhe-rs does not do any kind of check or interpretation it only stores it and passes it around